### PR TITLE
Phase 5: Uncertainty-Weighted Multi-Task Loss — GradNorm for Pressure vs Velocity (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model predicts 3 channels (p, Ux, Uy) with equal loss weighting, but our metric cares overwhelmingly about surface pressure. The current loss treats velocity and pressure equally, meaning velocity gradients may dominate training and dilute the pressure signal. Use uncertainty-weighted multi-task learning (Kendall et al., CVPR 2018) or GradNorm (Chen et al., ICML 2018) to automatically learn the optimal task weights that maximize pressure accuracy.

**Why this should work:**
- Pressure-first already helped by predicting p before v — this extends the same insight to the LOSS function
- In multi-task learning, naive equal weighting is almost never optimal. The tasks (Ux, Uy, p) have different scales, gradients, and learning dynamics.
- Uncertainty weighting learns a log-variance per task, naturally down-weighting easy tasks (velocity) and up-weighting hard tasks (pressure).
- GradNorm dynamically adjusts weights so all tasks train at equal rates.
- This is complementary to thorfinn's surface pressure weighting experiment (#1918) — that experiment weights surface vs volume, this one weights channels (p vs Ux vs Uy).

**References:**
- Kendall et al., "Multi-Task Learning Using Uncertainty to Weigh Losses" (CVPR 2018)
- Chen et al., "GradNorm: Gradient Normalization for Adaptive Loss Balancing" (ICML 2018)

## Instructions

See first comment for implementation details.

## Baseline (8-seed)
val/loss 0.404+/-0.004, p_in 13.33+/-0.58, p_oodc 8.37+/-0.22, p_tan 33.57+/-0.44, p_re 24.58+/-0.13

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT violet:

## Results: Uncertainty-Weighted Multi-Task Loss (8 experiments)

All experiments ran ~180 min, reaching 159-160 epochs. Same Transolver backbone.

### Baseline (8-seed from PR body)
| val/loss | p_in | p_oodc | p_tan | p_re |
|----------|------|--------|-------|------|
| 0.404±0.004 | 13.33±0.58 | 8.37±0.22 | 33.57±0.44 | 24.58±0.13 |

### Results

| GPU | Experiment | val/loss | p_in | p_oodc | p_tan | p_re |
|-----|-----------|----------|------|--------|-------|------|
| 0 | uw-equal (log_vars=[0,0,0]) | 0.663 | 14.3 | 9.9 | 35.9 | 25.4 |
| 1 | uw-pbias (log_vars=[0,0,-1]) | 0.683 | 13.8 | 10.6 | 36.2 | 25.6 |
| 2 | uw-pbias2 (log_vars=[0,0,-2]) | 0.694 | 14.5 | 10.5 | 36.0 | 25.6 |
| 3 | cpw2 (2x pressure) | 0.410 | 13.3 | 8.5 | 33.7 | 24.6 |
| 4 | **cpw3 (3x pressure)** | 0.409 | **12.6** | **8.4** | 33.9 | 24.6 |
| 5 | cpw5 (5x pressure) | 0.420 | 13.3 | 8.4 | 34.3 | 24.7 |
| 6 | **baseline s42** | **0.399** | 12.8 | 8.4 | **33.1** | 24.6 |
| 7 | baseline s43 | 0.404 | 13.0 | **8.3** | 33.9 | 24.6 |

### Command used (GPU 0 example)
```
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent violet --field_decoder --adaln_output --use_lion --lr 2e-4 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad --pressure_first --pressure_deep --seed 42 --uncertainty_weight --uw_init_p 0.0 --wandb_group phase5_multitask_loss --wandb_name "violet/uw-equal"
```

### Analysis

**Mixed result: Kendall uncertainty weighting hurts badly; fixed pressure upweight is marginal.**

1. **Uncertainty weighting (GPUs 0-2) is clearly harmful** — val/loss 0.66-0.69 vs baseline 0.40. The learned log-variance parameters cause the loss to diverge because the regularization term (0.5 * log_var) rewards increasing variance (reducing precision) for all channels. With only 3 parameters to learn, the optimization landscape is degenerate — the model learns to down-weight everything rather than selectively up-weight pressure.

2. **Fixed 3x pressure weight (GPU 4) is the best experimental variant** — p_in=12.6 vs baseline 12.8 (marginal), but val/loss 0.409 vs 0.399 (slightly worse). The improvement on in-dist surface pressure comes at the cost of worse OOD metrics and tandem transfer.

3. **Fixed 5x is too aggressive** — val/loss 0.420, worse than 2x or 3x. Over-weighting pressure causes velocity prediction to degrade.

4. **Fixed 2x is essentially baseline** — 0.410 val/loss, 13.3 p_in, within noise of baseline.

5. **The baseline already has effective pressure weighting** — the existing pressure-first + pressure-deep architecture already specializes the model for pressure prediction. Additional loss-level weighting provides diminishing returns.

**Root cause for uncertainty weighting failure:** The Kendall formulation `L/σ² + log(σ)` is designed for regression with MSE loss, but our training uses MAE (L1). With MAE, the optimal σ minimizes `L/σ² + log(σ)` which has a different equilibrium than with MSE. The formulation may need to be `L/σ + log(σ)` for L1 loss. Additionally, the per-channel surface loss already incorporates the existing surf_weight scaling, creating a double-weighting interaction.

**Recommendation:** The fixed 3x pressure weight shows a tiny p_in improvement (12.6 vs 12.8) but it's within seed variance (baseline range 12.95±0.58) and comes with worse val/loss. Not worth merging. The existing pressure-first architecture is already the best way to prioritize pressure.

W&B runs: `j15ay8gl`, `fshzexjc`, `6qg4sugk`, `oq8cx9yg`, `mjdz4hm8`, `n00rsatw`, `gbg4f38b`, `t6j77hxp`
